### ISSUE TITLE
Wrong segment url, download failed with error 404

### ISSUE THREAD

**Issue by mediaminister** [peak3d/inputstream.adaptive#707]
@matthuisman There is a regression introduced in https://github.com/xbmc/inputstream.adaptive/commit/10d7f6f6d60678a6e5009a2b72df7ab9bad986c9 that breaks playing uplynk.com streams:

MPEG-DASH manifest url: https://content.uplynk.com/channel/e11a05356cc44198977436418ad71832.mpd
MPEG-DASH manifest: [e11a05356cc44198977436418ad71832.mpd.txt](https://github.com/xbmc/inputstream.adaptive/files/6604789/e11a05356cc44198977436418ad71832.mpd.txt)

Kodi log
```
2021-06-06 21:29:13.692 T:5631    DEBUG <general>: AddOnLog: inputstream.adaptive: AdaptiveStream: trying to reload segment ...
2021-06-06 21:29:13.693 T:5631    DEBUG <general>: CurlFile::ParseAndCorrectUrl() adding custom header option 'connection: keep-alive'
2021-06-06 21:29:13.693 T:5631    DEBUG <general>: CurlFile::Open(0x7f7e24003220) https://content.uplynk.com/channel/F00000584.m4f?pbs=28b714541f8940cfba9601cfa09baa89&_jt=l&chid=e11a05356cc44198977436418ad71832&cloud=aws&si=0
2021-06-06 21:29:13.816 T:5631    ERROR <general>: AddOnLog: inputstream.adaptive: Download https://content.uplynk.com/channel/F00000584.m4f?pbs=28b714541f8940cfba9601cfa09baa89&_jt=l&chid=e11a05356cc44198977436418ad71832&cloud=aws&si=0 failed with error: 404
```

**Comment by matthuisman** [comment]
Can you get me a log from an older IA showing it working please. That will show the correct urls. I see it's using a Location tag.  Not sure if IA fully supports that

**Comment by mediaminister** [comment]
Yes, the stream plays perfect with https://github.com/xbmc/inputstream.adaptive/commit/ca527a688901c2abe671550552944491a43fcaa9

FAILED with 404: https://content.uplynk.com/channel/F00000506.m4f?pbs=8df803889ef9445fbc5feeab6792ba65&_jt=l&chid=e11a05356cc44198977436418ad71832&cloud=azure&si=0

RIGHT URL: https://x-default-stgec.uplynk.com/aeui/slices/51a/9578692592b347f6b5f6b7cbb206f368/51aaf401283c45de9aa6d7369c0910d7/F00000554.m4f?pbs=ae708a91fd064970a1497e62c1b1dcfa&_jt=l&chid=e11a05356cc44198977436418ad71832&cloud=aws&si=0

Working log:
```
2021-06-06 21:48:40.646 T:6185    DEBUG <general>: AddOnLog: inputstream.adaptive: OpenStream(1003)
2021-06-06 21:48:40.646 T:6185    DEBUG <general>: AddOnLog: inputstream.adaptive: Selecting stream with conditions: w: 0, h: 0, bw: 1514431
2021-06-06 21:48:40.646 T:6185    DEBUG <general>: CurlFile::ParseAndCorrectUrl() adding custom header option 'connection: keep-alive'
2021-06-06 21:48:40.646 T:6185    DEBUG <general>: CurlFile::Open(0x7f690c062250) https://x-default-stgec.uplynk.com/aeui/slices/51a/9578692592b347f6b5f6b7cbb206f368/51aaf401283c45de9aa6d7369c0910d7/TK_init.mp4?pbs=ae708a91fd064970a1497e62c1b1dcfa&_jt=l&chid=e11a05356cc44198977436418ad71832&cloud=aws&si=0
2021-06-06 21:48:40.662 T:6185    DEBUG <general>: AddOnLog: inputstream.adaptive: Download https://x-default-stgec.uplynk.com/aeui/slices/51a/9578692592b347f6b5f6b7cbb206f368/51aaf401283c45de9aa6d7369c0910d7/TK_init.mp4?pbs=ae708a91fd064970a1497e62c1b1dcfa&_jt=l&chid=e11a05356cc44198977436418ad71832&cloud=aws&si=0 finished, avg speed: 1890457.92byte/s, current speed: 45533.00byte/s
2021-06-06 21:48:40.663 T:6185    DEBUG <general>: AddOnLog: inputstream.adaptive: UpdateSampleDescription: codec 1836069985
2021-06-06 21:48:40.663 T:6185    DEBUG <general>: AddOnLog: inputstream.adaptive: GetStream(1003)
2021-06-06 21:48:40.663 T:6185    DEBUG <general>: CDVDDemuxClient::RequestStream(): added/updated stream 1003 with codec_id 86018
2021-06-06 21:48:40.663 T:6185     INFO <general>: Finding audio codec for: 86018
2021-06-06 21:48:40.665 T:6185     INFO <general>: CDVDAudioCodecFFmpeg::Open() Successful opened audio decoder aac
2021-06-06 21:48:40.665 T:6185     INFO <general>: Creating audio thread
2021-06-06 21:48:40.665 T:6197    DEBUG <general>: Thread VideoPlayerAudio start, auto delete: false
2021-06-06 21:48:40.665 T:6197     INFO <general>: running thread: CVideoPlayerAudio::Process()
2021-06-06 21:48:40.665 T:6197    DEBUG <general>: CVideoPlayerAudio - CDVDMsg::GENERAL_PAUSE: 0
2021-06-06 21:48:40.665 T:6185    DEBUG <general>: CVideoPlayer::SetCaching - caching state 2
2021-06-06 21:48:40.665 T:6185    DEBUG <general>: CDVDClock::SetSpeedAdjust - adjusted:0.000000
2021-06-06 21:48:40.665 T:6185    DEBUG <general>: CVideoPlayer::HandleMessages - player 2 reported state: 0
2021-06-06 21:48:40.665 T:6185    DEBUG <general>: CVideoPlayer::HandleMessages - player 1 reported state: 0
2021-06-06 21:48:40.665 T:6185    DEBUG <general>: CDVDClock::SetSpeedAdjust - adjusted:-0.050000
2021-06-06 21:48:40.665 T:6150    DEBUG <general>: OnAVChange: CApplication::OnAVChange
2021-06-06 21:48:40.665 T:6192    DEBUG <general>: CurlFile::ParseAndCorrectUrl() adding custom header option 'connection: keep-alive'
2021-06-06 21:48:40.665 T:6192    DEBUG <general>: CurlFile::Open(0x7f68fc003770) https://x-default-stgec.uplynk.com/aeui/slices/51a/9578692592b347f6b5f6b7cbb206f368/51aaf401283c45de9aa6d7369c0910d7/F00000554.m4f?pbs=ae708a91fd064970a1497e62c1b1dcfa&_jt=l&chid=e11a05356cc44198977436418ad71832&cloud=aws&si=0
2021-06-06 21:48:40.675 T:6197    DEBUG <general>: CDVDAudio::Pause - pausing audio stream
2021-06-06 21:48:40.795 T:6196    DEBUG <general>: CurlFile::ParseAndCorrectUrl() adding custom header option 'connection: keep-alive'
2021-06-06 21:48:40.795 T:6196    DEBUG <general>: CurlFile::Open(0x7f68f4003760) https://x-default-stgec.uplynk.com/aeui/slices/51a/9578692592b347f6b5f6b7cbb206f368/51aaf401283c45de9aa6d7369c0910d7/TK_00000554.m4f?pbs=ae708a91fd064970a1497e62c1b1dcfa&_jt=l&chid=e11a05356cc44198977436418ad71832&cloud=aws&si=0
2021-06-06 21:48:40.795 T:6196    DEBUG <general>: easy_acquire - Created session to https://x-default-stgec.uplynk.com
2021-06-06 21:48:40.893 T:6192    DEBUG <general>: AddOnLog: inputstream.adaptive: Download https://x-default-stgec.uplynk.com/aeui/slices/51a/9578692592b347f6b5f6b7cbb206f368/51aaf401283c45de9aa6d7369c0910d7/F00000554.m4f?pbs=ae708a91fd064970a1497e62c1b1dcfa&_jt=l&chid=e11a05356cc44198977436418ad71832&cloud=aws&si=0 finished, avg speed: 2116368.55byte/s, current speed: 2336991.00byte/s
```

**Comment by matthuisman** [comment]
Thanks this is the pr for those changes:
https://github.com/xbmc/inputstream.adaptive/pull/668

Looking at the manifest it looks like baseurl comes after the segment template on the video set. So when IA is calculating the init and segment template it hasn't yet set the base url so calculates them off the previous base url (or manifest url if no baseurl)

Now it used to work by luck as IA was just using the previous adaption set init url (which luckily uses full urls) to work out the baseurl. Which is incorrect as per the PR.

I've found other occasions where the order of elements affects IA. So we going to need a bit of a think about this one. We can't revert those changes as that breaks other streams and is wrong @glennguy . Hmmm? We almost need to only calculate full ulrs after all the manifest has been read. We eventually need this anyway to correctly support multiple base urls. The full urls should be calculated at the last possible time (just before downloading)

**Comment by mediaminister** [comment]
No, BaseURL comes before SegmentTemplate.

**Comment by matthuisman** [comment]
Not on the 2nd ,set.
Wait. Maybe it does. I'm on phone so hard to see formatting. Let me get on a pc

[e11a05356cc44198977436418ad71832.pretty.txt](https://github.com/xbmc/inputstream.adaptive/files/6604866/e11a05356cc44198977436418ad71832.pretty.txt)

OK, it is after... however - it was still luck that made it originally work as it was using the initialization url to get the base url.
If you change the init url to be relative - i think you'd find it won't then work on older IA as well.
I can see why now maybe peak was doing that. assuming media is always found in same location as init.
but that was breaking adaption sets using relative to the higher up base url

**So we just need to see why IA is not using the base url element found earlier in the adaption set.**
again - this will break if baseurl comes after SegmentTemplate though but maybe will do for now?

it looks like we are just ignoring the base url here:
https://github.com/xbmc/inputstream.adaptive/blob/Matrix/src/parser/DASHTree.cpp#L602

I'm suspecting IA doesn't yet support base url in representations. Shouldn't be too hard to add. By default it will be set to the adaption set base url. And then updated if base url in the rep. Then the segment calc code will use this new rep base url instead of the adaption set base url

**Comment by matthuisman** [comment]
@mediaminister 
could you please try a build from here and confirm that fixes it
https://jenkins.kodi.tv/blue/organizations/jenkins/xbmc%2Finputstream.adaptive/detail/PR-708/1/artifacts

**Comment by mediaminister** [comment]
Yes, you fixed it. Many thanks!


```
2021-06-07 11:18:50.842 T:29041   DEBUG <general>: OnAVChange: CApplication::OnAVChange
2021-06-07 11:18:50.842 T:29095   DEBUG <general>: CurlFile::ParseAndCorrectUrl() adding custom header option 'connection: keep-alive'
2021-06-07 11:18:50.842 T:29095   DEBUG <general>: CurlFile::Open(0x7ff390003770) https://x-default-stgec.uplynk.com/aeui/slices/faa/9578692592b347f6b5f6b7cbb206f368/faa031989abd401d917829efb1cc8c38/F00000754.m4f?pbs=6c191fc49e284385ac289380e09f8c42&_jt=l&chid=e11a05356cc44198977436418ad71832&cloud=aws&si=0
2021-06-07 11:18:50.852 T:29100   DEBUG <general>: CVideoPlayerAudio - CDVDMsg::GENERAL_PAUSE: 0
2021-06-07 11:18:50.852 T:29100   DEBUG <general>: CDVDAudio::Pause - pausing audio stream
2021-06-07 11:18:50.866 T:29027   DEBUG <general>: Inhibiting OS screen saver
2021-06-07 11:18:50.867 T:29101   DEBUG <general>: Thread Timer start, auto delete: false
2021-06-07 11:18:51.012 T:29099   DEBUG <general>: CurlFile::ParseAndCorrectUrl() adding custom header option 'connection: keep-alive'
2021-06-07 11:18:51.012 T:29099   DEBUG <general>: CurlFile::Open(0x7ff388003760) https://x-default-stgec.uplynk.com/aeui/slices/faa/9578692592b347f6b5f6b7cbb206f368/faa031989abd401d917829efb1cc8c38/TK_00000754.m4f?pbs=6c191fc49e284385ac289380e09f8c42&_jt=l&chid=e11a05356cc44198977436418ad71832&cloud=aws&si=0
2021-06-07 11:18:51.012 T:29099   DEBUG <general>: easy_acquire - Created session to https://x-default-stgec.uplynk.com
2021-06-07 11:18:51.068 T:29095   DEBUG <general>: AddOnLog: inputstream.adaptive: Download finished: https://x-default-stgec.uplynk.com/aeui/slices/faa/9578692592b347f6b5f6b7cbb206f368/faa031989abd401d917829efb1cc8c38/F00000754.m4f?pbs=6c191fc49e284385ac289380e09f8c42&_jt=l&chid=e11a05356cc44198977436418ad71832&cloud=aws&si=0 , avg speed: 2711806.12byte/s, current speed: 2415862.00byte/s

```

**Comment by glennguy** [comment]
Fixed in #708
